thorlcr/activities/hashdistrib/thhashdistribslave.cpp

```diff
@@ -3737,119 +3737,341 @@ class HashJoinSlaveActivity : public CSlaveActivity, implements IStopInput
-    activity.ActPrintLog("%s", str.append(globalAggTable->elementCount()).append("entries").str());
-    return globalAggTable.getClear();
+IRowStream *mergeLocalAggs(Owned<IHashDistributor> &distributor, CActivityBase &activity, IHThorRowAggregator &helper, IHThorHashAggregateExtra &helperExtra, IRowStream *localAggStream, mptag_t mptag)
+{
+    Owned<IRowStream> strm;
+    ICompare *elementComparer = helperExtra.queryCompareElements();
+    if (!distributor)
+        distributor.setown(createPullHashDistributor(&activity, activity.queryContainer().queryJobChannel().queryJobComm(), mptag, false, NULL, "MERGEAGGS"));
+    Owned<IThorRowInterfaces> rowIf = activity.getRowInterfaces(); // create new rowIF / avoid using activities IRowInterface, otherwise suffer from circular link
+    IHash *hasher = helperExtra.queryHashElement();
+    strm.setown(distributor->connect(rowIf, localAggStream, hasher, elementComparer));
+    IEngineRowAllocator *rowAllocator = activity.queryRowAllocator();
+
+    class CAggregatingStream : public CSimpleInterfaceOf<IRowStream>
+    {
+        size32_t sz = 0;
+        IEngineRowAllocator &rowAllocator;
+        RtlDynamicRowBuilder rowBuilder;
+        Owned<IRowStream> input;
+        ICompare &cmp;
+        IHThorRowAggregator &helper;
+        IHashDistributor &distributor;
+    public:
+        CAggregatingStream(IHThorRowAggregator &_helper, IEngineRowAllocator &_rowAllocator, ICompare &_cmp, IHashDistributor &_distributor)
+            : helper(_helper), rowAllocator(_rowAllocator), cmp(_cmp), distributor(_distributor), rowBuilder(_rowAllocator)
+        {
+        }
+        void start(IRowStream *_input)
+        {
+            input.setown(_input);
+        }
+        // IRowStream
+        virtual const void *nextRow() override
+        {
+            for (;;)
+            {
+                OwnedConstThorRow row = input->nextRow();
+                if (!row)
+                {
+                    if (sz)
+                    {
+                        const void *row = rowBuilder.finalizeRowClear(sz);
+                        sz = 0;
+                        return row;
+                    }
+                    return nullptr;
+                }
+                else if (sz)
+                {
+                    if (0 == cmp.docompare(row, rowBuilder.getUnfinalized()))
+                        sz = helper.mergeAggregate(rowBuilder, row);
+                    else
+                    {
+                        const void *ret = rowBuilder.finalizeRowClear(sz);
+                        sz = cloneRow(rowBuilder, row, rowAllocator.queryOutputMeta());
+                        return ret;
+                    }
+                }
+                else
+                    sz = cloneRow(rowBuilder, row, rowAllocator.queryOutputMeta());
+            }
+            return nullptr;
+        }
+        virtual void stop() override
+        {
+            sz = 0;
```

Comment:
ok will add, though in practice as it stands it wouldn't make any difference as object isn't reused.